File: LeetCode/path_sum.rb
#  Given a binary tree and a sum, determine if the tree has a root-to-leaf path such that adding up all the values along the path equals the given sum.
#  For example:
#  Given the below binary tree and sum = 22,
#
#                5
#               / \
#              4   8
#             /   / \
#            11  13  4
#           /  \      \
#          7    2      1
#
#  Return true, as there exist a root-to-leaf path 5->4->11->2 which sum is 22.
#

class Node < Struct.new(:value, :left, :right)
end

root = Node.new(5,
                Node.new(4,
                         Node.new(11,
                                  Node.new(7,
                                           nil,
                                           nil),
                                  Node.new(2,
                                           nil,
                                           nil)),
                        nil),
                Node.new(8,
                         Node.new(13,
                                  nil,
                                  nil),
                         Node.new(4,
                                  nil,
                                  Node.new(1,
                                           nil,
                                           nil))))

stack = []
current = root
while stack.any? || current
    if current.nil?
        current = stack.pop
        print "#{current.value} "
        current = current.right
    else
        stack.push(current)
        current = current.left
    end
end

puts

def path_sum(node, sum, target)
    return false if node.nil?
    if node.left.nil? && node.right.nil? 
        return sum + node.value == target
    end

    return path_sum(node.left, sum + node.value, target) || path_sum(node.right, sum + node.value, target) 
end

puts path_sum(root, 0, 30)
puts path_sum(root, 0, 22)
puts path_sum(root, 0, 26)

File: LeetCode/path_sum.cs
using System;
using System.Collections.Generic;

class Node {
    public int Value { get; set; }
    public Node Left { get; set; }
    public Node Right { get; set; }
}

class Program {

    static bool Sum(Node current, int target, int sum, int[] prev) {
        if (current.Left == null && current.Right == null) {
            if (target == sum + current.Value) {
                int node = current.Value;
                while (0 < node && node < prev.Length) {
                    Console.Write("{0} ", node);
                    node = prev[node];
                }

                Console.WriteLine();
                return true;
            } else {
                return false;
            }
        }
        
        bool left = false;
        bool right = false;
        if (current.Left != null) {
            prev[current.Left.Value] = current.Value;
            left = Sum(current.Left, target, sum + current.Value, prev);
        }

        if (current.Right!= null) {
            prev[current.Right.Value] = current.Value;
            right = Sum(current.Right, target, sum + current.Value, prev);
        }

        return left || right;
    }

    static void Main() {
        Node root = new Node {
                                Value = 5,
                                Left = new Node {
                                                    Value = 4,
                                                    Left = new Node {
                                                                        Value = 11,
                                                                        Left = new Node {
                                                                                            Value = 7,
                                                                                            Left = null,
                                                                                            Right = null
                                                                                        },
                                                                        Right = new Node {
                                                                                            Value = 2,
                                                                                            Left = null,
                                                                                            Right = null
                                                                                         }
                                                                    },
                                                    Right = null
                                                },
                                Right = new Node {
                                                    Value = 8,
                                                    Left = new Node {
                                                                        Value = 13,
                                                                        Left = null,
                                                                        Right = null
                                                                    },
                                                    Right = new Node {
                                                                        Value = 6,
                                                                        Left = null,
                                                                        Right = new Node {
                                                                                            Value = 1,
                                                                                            Left = null,
                                                                                            Right = null
                                                                                         }
                                                                     }
                                                 }
                             };

        Console.WriteLine(Sum(root, 22, 0, new int[100]));
    }
}
